### DISCUSSION TITLE
Possible temporary files issue

### DISCUSSION THREAD

**Discussion by alexjnoble** [nextpyp/.github#11] [Q&A] [answered]
Hello,

I am working through the [nextPYP tomography tutorial](https://nextpyp.app/files/pyp/0.5.1/docs/tutorials/tomo_empiar_10164.html) and when I run Step 6: 

Reference-based refinement job, it eventually ends with a temporary file not found error:

`2023-11-17 13:47:47 [ERROR] pyp/system/local_run.py:72 | /bin/sh: line 0: cd: /common/sw/nextpyp/tmp/1262_1/TS_01: No such file or directory`

See attached image and log from one of the jobs as well. This is running on a SLURM cluster with 400+ GB free. This issue happened for each of the 5 SLURM jobs that were launched by nextPYP. Our IT says that our nextpyp /tmp/ file directories are not being deleted outside of nextPYP, so this may be a nextPYP bug.

Thank you,
-Alex

[nextpyp_testing6.txt](https://github.com/nextpyp/.github/files/13418600/nextpyp_testing6.txt)
![nextpyp_testing6](https://github.com/nextpyp/.github/assets/30907567/b6d158c6-b6e0-4be6-9bdd-46dbdfc14f15)

**Comment by abartesaghi** [comment, marked as answer]
Alex, nextPYP should be handling temporary storage properly. We've tested this extensively and haven't seen issues like the one you are noticing. It looks as if the `/common/sw/nextpyp/tmp/1262_1/TS_01` folder is being deleted by an external process somehow, though that doesn't seem to be happening according to your IT people. We noticed that your processes ran for several hours (over a day?) before failing so I'm wondering if there are any "cleaning" processes running on that time scale.

Also, the scratch folder setting (`pyp.scratch` which in your case is set to `/common/sw/nextpyp/tmp`) is meant to be local to the nodes, ideally fast SSDs. If a network share is used instead, it is important to keep in mind that you will need to have enough space to fit the particle stacks for the entire dataset. For example, the tomography tutorial uses about 30k particles, so assuming 41 tilts per particle and a box size of 192x192 pixels (at least for step 6), this will amount to 170GB worth of particle stacks plus storage for the raw tilt-series and other volumes. That is to say that if other nextPYP processes are running on that same location, it could quickly fill up the +400GB.

If you don't have access to local scratch on the compute nodes or need to use shared network space for some reason, one way to reduce storage requirements is to process tilt-series serially (instead of parallel), this can be achieved using a `Bundle size` (`Resources` tab) equal to the number of tilt-series in your dataset. You could try running in this mode to rule out any issues with storage space.

**Comment by alexjnoble** [comment]
Thanks for the reply, Alberto. I'm trying your Bundle size idea. I'll report back what happens.

-Alex

**Comment by alexjnoble** [comment]
Hello,

I changed the Bundle size to 5, which is the number of tilt-series in the tomography tutorial dataset (I'm still on Step 6). The job has been running for about 24 hours, but it says that it will take 436 hours longer. We are using a Slurm cluster and there are 5 nodes available. nextPYP submitted 1 job that is using 10 threads. I have attached several screenshots of the streaming log, the Resources configuration for the job, and 'top' running on the node that nextPYP is using. I must be configuring something wrong. Could you let me know what should be changed?

Thanks,
-Alex

![nextpyp_testing7a](https://github.com/nextpyp/.github/assets/30907567/fe697ccc-355c-4d0a-8d82-ed6d6f469041)
![nextpyp_testing7b](https://github.com/nextpyp/.github/assets/30907567/6d565514-11ed-48c0-8872-2e0e91e9909c)
![nextpyp_testing7c](https://github.com/nextpyp/.github/assets/30907567/bdc488dd-bdf7-4984-a8e6-c291e1392fb0)
![nextpyp_testing7e](https://github.com/nextpyp/.github/assets/30907567/508d444c-99ef-4935-a2ee-01b8a0025057)

**Comment by abartesaghi** [comment]
Alex, the reason I suggested increasing the bundle size was to reduce the overall storage requirements, but it does mean that you will run the 5 tilt-series serially, which will slow you down by a factor of 5. My first suggestion would be to make sure you are using storage that is local to the compute node as scratch space. This has to be configured in your instance as described [here](https://nextpyp.app/files/pyp/0.5.2/docs/install/install-web.html#step-5-recommended-configure-access-to-system-resources). This will do two things: it will significantly speed up all nextPYP processes, and you will not have to deal with the storage issues you mentioned earlier. Looking at the logs, it looks like you are using network storage as local scratch (`igloo.semc.nysbc.org:/common/sw/nextpyp/tmp`), so this should help. Second, make sure you are using as many `Threads per task` (`Resources` tab) as cores/hyper-cores you have available in the compute nodes (you may need to adjust the `Memory per task` accordingly, roughly 4-8GB per thread depending on how much RAM you have available). If you set the `Bundle size` back to 1, you will be able to run the 5 tilt-series in parallel which will cut runtime by 5x (assuming you have access to 5 nodes).

**Reply by abartesaghi** [reply]
Since we haven't heard back from you, we assume this issue was resolved, but please let us know otherwise.

**Comment by alexjnoble** [comment]
Hi Alberto,

I haven't solved this yet - we'll hopefully be getting some CPU nodes up on our cluster. I haven't been able to complete a particle tilt-series refinement job yet. We're still working on it...

-Alex

**Reply by abartesaghi** [reply]
OK, let us know if you are getting any errors or need help setting up the runs.
